src/libos/src/vm/vm_manager.rs

```diff
@@ -440,31 +480,68 @@ impl VMManager {
+                        if let Some((backed_file, offset)) = writeback_file {
+                            // Write back file should start from new offset
+                            writeback_file =
+                                Some((backed_file.clone(), offset + containing_vma.size()));
+                            VMInitializer::LoadFromFile {
+                                file: backed_file,
+                                offset: offset + containing_vma.size(), // file-backed mremap should start from the end of previous mmap/mremap file
+                            }
+                        } else {
+                            VMInitializer::FillZeros()
+                        };
                     let mmap_ops = VMMapOptionsBuilder::default()
                         .size(prefered_new_range.size())
                         .addr(VMMapAddr::Need(prefered_new_range.start()))
                         .perms(perms)
-                        .initializer(VMInitializer::FillZeros())
+                        .initializer(vm_initializer_for_new_range)
+                        .writeback_file(writeback_file)
                         .build()?;
                     (Some(mmap_ops), Some(old_addr))
                 } else {
+                    // Need to move old range to a new range and init the new range
+                    let vm_initializer_for_new_range =
+                        if let Some((backed_file, offset)) = &writeback_file {
+                            VMInitializer::CopyOldAndReadNew {
+                                old_range: old_range,
+                                file: backed_file.clone(),
+                                offset: offset + containing_vma.size(),
+                            }
+                        } else {
+                            VMInitializer::CopyFrom { range: old_range }
+                        };
                     let mmap_ops = VMMapOptionsBuilder::default()
                         .size(new_size)
                         .addr(VMMapAddr::Any)
                         .perms(perms)
-                        .initializer(VMInitializer::CopyFrom { range: old_range })
+                        .initializer(vm_initializer_for_new_range)
+                        .writeback_file(writeback_file)
                         .build()?;
                     // Cannot determine the returned address for now, which can only be obtained after calling mmap
                     let ret_addr = None;
                     (Some(mmap_ops), ret_addr)
                 }
             }
             (MRemapFlags::FixedAddr(new_addr), _) => {
+                let vm_initializer_for_new_range =
+                    if let Some((backed_file, offset)) = &writeback_file {
+                        VMInitializer::CopyOldAndReadNew {
+                            old_range: old_range,
+                            file: backed_file.clone(),
+                            offset: offset + containing_vma.size(),
```

Comment:
The calculation of offset has the same problem as pointed out in the comments above.